### ISSUE TITLE
[BUG] Mojo destroys value too eagerly

### ISSUE THREAD

**Issue by soraros** [modularml/mojo#3473]
### Bug description

As title. I think it's related to `VariadicPack`.
CC @lattner.

### Steps to reproduce

```mojo
fn main():
  l = List[String]("a", "b", "c")
  print(l[0], len(l))  # prints: ` 3`
```

### System information

```shell
Mojo 2024.9.1205 (e6c1b2aa) on Docker, Intel Mac
```

**Comment by ematejska** [comment]
For mojo 2024.9.1605 (900128df), reproduces.  I see output of "( 3"

**Comment by ematejska** [comment]
Doesn't reproduce with mojo 24.5.0 (e8aacb95).

**Comment by soraros** [comment]
Seems to be fixed on mojo 2024.9.2405 (094f5635).

**Comment by lattner** [comment]
I just verified it still works with main, but I don't think it is related to the AnyLifetime changes:

```
> mojo test.mojo
a 3
```
